File: x8086NetEmu/Adapters/Serial/MouseAdapter.vb
﻿Public Class MouseAdapter
    Inherits Adapter
    Implements IExternalInputHandler

    Private Class SerialMouse
        Public reg(8 - 1) As UInteger
        Public buf(16 - 1) As UInteger
        Public bufPtr As UInteger
    End Class

    Private mCPU As X8086
    Private sm As New SerialMouse()
    Private irq As PIC8259.IRQLine

    Private lastX As Integer = Integer.MinValue
    Private lastY As Integer = Integer.MinValue

    Private Const M As Integer = Asc("M")

    Public Sub New(cpu As X8086)
        mCPU = cpu
        If mCPU.PIC IsNot Nothing Then irq = mCPU.PIC.GetIrqLine(4)

        For i As UInteger = &H3F8 To &H3F8 + 7
            ValidPortAddress.Add(i)
        Next
    End Sub

    Public Overrides Function [In](port As UInteger) As UInteger
        Dim tmp As Integer

        Select Case port
            Case &H3F8 ' Transmit/Receive Buffer
                If irq IsNot Nothing Then
                    tmp = sm.buf(0)
                    Array.Copy(sm.buf, 1, sm.buf, 0, 15)
                    sm.bufPtr = (sm.bufPtr - 1) And &HF

                    If sm.bufPtr < 0 Then sm.bufPtr = 0
                    If sm.bufPtr > 0 Then irq.Raise(True)

                    sm.reg(4) = (Not sm.reg(4)) And 1
                End If

                Return tmp
            Case &H3FD ' Line Status Register - LSR
                If sm.bufPtr > 0 Then
                    tmp = 1
                Else
                    tmp = 0
                End If

                'Return tmp
                'Return &H60 Or tmp
                Return &H1
        End Select

        Return sm.reg(port And 7)
    End Function

    Public Overrides Sub Out(port As UInteger, value As UInteger)
        Dim oldReg As Integer = sm.reg(port And 7)
        sm.reg(port And 7) = value

        Select Case port
            Case &H3FC ' Modem Control Register - MCR
                If (value And 1) <> (oldReg And 1) Then ' Software toggling of this register
                    sm.bufPtr = 0 '                       causes the mouse to reset and fill the buffer
                    BufSerMouseData(M) '                  with a bunch of ASCII 'M' characters.
                    BufSerMouseData(M) '                  this is intended to be a way for
                    BufSerMouseData(M) '                  drivers to verify that there is
                    BufSerMouseData(M) '                  actually a mouse connected to the port.
                    BufSerMouseData(M)
                End If
                mCPU.Flags.OF = 1
        End Select
    End Sub

    Private Sub BufSerMouseData(value As Byte)
        If irq IsNot Nothing Then
            If sm.bufPtr = 16 Then Exit Sub
            If sm.bufPtr = 0 Then irq.Raise(True)

            sm.buf(sm.bufPtr) = value
            sm.bufPtr += 1
        End If
    End Sub

    Public Sub HandleInput(e As ExternalInputEvent) Implements IExternalInputHandler.HandleInput
        Dim m As MouseEventArgs = CType(e.TheEvent, MouseEventArgs)

        ' TODO: The scaling code needs to be adjusted so that the size of the screen is properly mapped to the emulator window

        If lastX = Integer.MinValue Then lastX = m.X
        If lastY = Integer.MinValue Then lastY = m.Y

        Dim va As VideoAdapter = mCPU.VideoAdapter
        Dim rc As Control

        If TypeOf va Is CGAWinForms Then
            rc = CType(va, CGAWinForms).RenderControl
        Else
            rc = CType(va, VGAWinForms).RenderControl
        End If

        Dim rw As Double = rc.Width / If(va.MainMode = VideoAdapter.MainModes.Text, rc.Width, va.GraphicsResolution.Width)
        Dim rh As Double = rc.Height / If(va.MainMode = VideoAdapter.MainModes.Text, rc.Height, va.GraphicsResolution.Height)

        Dim rX As Integer = 1
        Dim ry As Integer = 1

        Dim p As New Point(m.X / va.Zoom, m.Y / va.Zoom / (rh / rw))

        Dim sX As Integer = Math.Sign(p.X - lastX)
        Dim sy As Integer = Math.Sign(p.Y - lastY)

        Do
            rX = If(p.X = lastX, 0, sX)
            ry = If(p.Y = lastY, 0, sy)

            Dim highbits As Integer = 0
            If rX < 0 Then highbits = 3
            If ry < 0 Then highbits = highbits Or &HC

            Dim btns As Integer = 0
            If (m.Button And MouseButtons.Left) = MouseButtons.Left Then btns = btns Or 2
            If (m.Button And MouseButtons.Right) = MouseButtons.Right Then btns = btns Or 1

            BufSerMouseData(&H40 Or (btns << 4) Or highbits)
            BufSerMouseData(rX And &H3F)
            BufSerMouseData(ry And &H3F)

            lastX += rX
            lastY += ry
        Loop Until p.X = lastX AndAlso p.Y = lastY
    End Sub

    Public Overrides Sub CloseAdapter()

    End Sub

    Public Overrides Sub InitiAdapter()

    End Sub

    Public Overrides Sub Run()

    End Sub

    Public Overrides ReadOnly Property Name As String
        Get
            Return "Mouse"
        End Get
    End Property

    Public Overrides ReadOnly Property Description As String
        Get
            Return "Serial Mouse at COM1"
        End Get
    End Property

    Public Overrides ReadOnly Property Type As Adapter.AdapterType
        Get
            Return AdapterType.SerialMouseCOM1
        End Get
    End Property

    Public Overrides ReadOnly Property Vendor As String
        Get
            Return "xFX JumpStart"
        End Get
    End Property

    Public Overrides ReadOnly Property VersionMajor As Integer
        Get
            Return 0
        End Get
    End Property

    Public Overrides ReadOnly Property VersionMinor As Integer
        Get
            Return 0
        End Get
    End Property

    Public Overrides ReadOnly Property VersionRevision As Integer
        Get
            Return 1
        End Get
    End Property
End Class

File: x8086SharpEmu/Adapters/Serial/MouseAdapter.cs
using System.Collections.Generic;
using System;
using System.Linq;
using System.Drawing;
using System.Diagnostics;
using System.Xml.Linq;
using System.Collections;


using x8086SharpEmu;
using Assets.CCC.x8086Sharp.UnityHelpers;
using System.Windows.Forms;

namespace x8086SharpEmu
{
    public class MouseAdapter : Adapter, IExternalInputHandler
    {

        private class SerialMouse
        {
            public uint[] reg = new uint[8];
            public uint[] buf = new uint[16];
            public uint bufPtr;
        }

        private SerialMouse sm = new SerialMouse();
        private PIC8259.IRQLine irq;

        public Point MidPoint { get; set; }
        public bool IsCaptured { get; set; }

        //private const int M = Strings.Asc("M");
        private const int M = 77; //M as ascii

        public MouseAdapter(X8086 cpu) : base(cpu)
        {
            if (base.CPU.PIC != null)
            {
                irq = base.CPU.PIC.GetIrqLine((byte)4);
            }

            for (uint i = 0x3F8; i <= 0x3F8 + 7; i++)
            {
                ValidPortAddress.Add(i);
            }
        }

        public override ushort In(uint port)
        {
            int tmp = 0;

            if (port == ((uint)(0x3F8))) // Transmit/Receive Buffer
            {
                if (irq != null)
                {
                    tmp = (int)(sm.buf[0]);
                    Array.Copy(sm.buf, 1, sm.buf, 0, 15);
                    sm.bufPtr = (uint)((sm.bufPtr - 1) & 0xF);

                    if (sm.bufPtr < 0)
                    {
                        sm.bufPtr = (uint)0;
                    }
                    if (sm.bufPtr > 0)
                    {
                        irq.Raise(true);
                    }

                    sm.reg[4] = (uint)((~sm.reg[4]) & 1);
                }

                return (ushort)tmp;
            } // Line Status Register - LSR
            else if (port == ((uint)(0x3FD)))
            {
                tmp = (int)(sm.bufPtr > 0 ? 1 : 0);

                return (ushort)tmp;
                //Return &H60 Or tmp
                //Return &H1
            }

            return (ushort)(sm.reg[port & 7]);
        }

        public override void Out(uint port, ushort value)
        {
            int oldReg = (int)(sm.reg[port & 7]);
            sm.reg[port & 7] = (uint)(value);

            if (port == ((uint)(0x3FC))) // Modem Control Register - MCR
            {
                if ((value & 1) != (oldReg & 1)) // Software toggling of this register
                {
                    sm.bufPtr = (uint)0; //                       causes the mouse to reset and fill the buffer
                    BufSerMouseData((byte)M); //                  with a bunch of ASCII 'M' characters.
                    BufSerMouseData((byte)M); //                  this is intended to be a way for
                    BufSerMouseData((byte)M); //                  drivers to verify that there is
                    BufSerMouseData((byte)M); //                  actually a mouse connected to the port.
                    BufSerMouseData((byte)M);
                }
                base.CPU.Flags.OF = (byte)1;
            }
        }

        private void BufSerMouseData(byte value)
        {
            if (irq != null)
            {
                if (sm.bufPtr == 16)
                {
                    return;
                }
                if (sm.bufPtr == 0)
                {
                    irq.Raise(true);
                }

                sm.buf[sm.bufPtr] = (uint)(value);
                sm.bufPtr++;
            }
        }

        public void HandleInput(ExternalInputEvent e)
        {
            MouseEventArgs m = (MouseEventArgs)e.TheEvent;

            Point p = new Point(m.X - MidPoint.X, m.Y - MidPoint.Y);
            //If p.X <> 0 Then If p.X > 0 Then p.X = 1 Else p.X = -1
            //If p.Y <> 0 Then If p.Y > 0 Then p.Y = 1 Else p.Y = -1

            p.X = (int)(Math.Ceiling((double)Math.Abs(p.X) / 5) * Math.Sign(p.X));
            p.Y = (int)(Math.Ceiling((double)Math.Abs(p.Y) / 5) * Math.Sign(p.Y));

            byte highbits = (byte)0;
            if (p.X < 0)
            {
                highbits = (byte)3;
            }
            if (p.Y < 0)
            {
                highbits = (byte)(highbits | 0xC);
            }

            byte btns = (byte)0;
            if (((int)(m.Button) & (int)MouseButtons.Left) == (int)MouseButtons.Left)
            {
                btns = (byte)(btns | 2);
            }
            if (((int)(m.Button) & (int)MouseButtons.Right) == (int)MouseButtons.Right)
            {
                btns = (byte)(btns | 1);
            }

            BufSerMouseData((byte)(0x40 | (btns << 4) | highbits));
            BufSerMouseData((byte)(p.X & 0x3F));
            BufSerMouseData((byte)(p.Y & 0x3F));
        }

        public override void CloseAdapter()
        {

        }

        public override void InitiAdapter()
        {

        }

        public override void Run()
        {

        }

        public override string Name
        {
            get
            {
                return "Mouse";
            }
        }

        public override string Description
        {
            get
            {
                return "Serial Mouse at COM1";
            }
        }

        public override Adapter.AdapterType Type
        {
            get
            {
                return AdapterType.SerialMouseCOM1;
            }
        }

        public override string Vendor
        {
            get
            {
                return "xFX JumpStart";
            }
        }

        public override int VersionMajor
        {
            get
            {
                return 1;
            }
        }

        public override int VersionMinor
        {
            get
            {
                return 0;
            }
        }

        public override int VersionRevision
        {
            get
            {
                return 0;
            }
        }
    }
}
